cpp/include/ucxx/request_am.h

```diff
@@ -24,128 +25,69 @@ class RequestAm : public Request {
+   * available via the `getRecvBuffer()` method if the receive transfer request completed
+   * successfully.
    *
    * @throws ucxx::Error  if `endpoint` is not a valid
    *                      `std::shared_ptr<ucxx::Endpoint>`.
    *
    * @param[in] endpoint            the parent endpoint.
-   * @param[in] buffer              a raw pointer to the data to be transferred.
-   * @param[in] length              the size in bytes of the tag message to be transferred.
-   * @param[in] memoryType          the memory type of the buffer.
+   * @param[in] requestData         container of the specified message type, including all
+   *                                type-specific data.
    * @param[in] enablePythonFuture  whether a python future should be created and
    *                                subsequently notified.
    * @param[in] callbackFunction    user-defined callback function to call upon completion.
    * @param[in] callbackData        user-defined data to pass to the `callbackFunction`.
    *
    * @returns The `shared_ptr<ucxx::RequestAm>` object
    */
-  friend std::shared_ptr<RequestAm> createRequestAmSend(
-    std::shared_ptr<Endpoint> endpoint,
-    void* buffer,
-    size_t length,
-    ucs_memory_type_t memoryType,
-    const bool enablePythonFuture,
-    RequestCallbackUserFunction callbackFunction,
-    RequestCallbackUserData callbackData);
-
-  /**
-   * @brief Constructor for `std::shared_ptr<ucxx::RequestAm>` receive.
-   *
-   * The constructor for a `std::shared_ptr<ucxx::RequestAm>` object, creating a receive
-   * active message request, returning a pointer to a request object that can be later
-   * awaited and checked for errors. This is a non-blocking operation, and the status of
-   * the transfer must be verified from the resulting request object before the data can be
-   * consumed, the data is available via the `getRecvBuffer()` method if the transfer
-   * completed successfully.
-   *
-   * @throws ucxx::Error  if `endpoint` is not a valid
-   *                      `std::shared_ptr<ucxx::Endpoint>`.
-   *
-   * @param[in] endpoint            the parent endpoint.
-   * @param[in] enablePythonFuture  whether a python future should be created and
-   *                                subsequently notified.
-   * @param[in] callbackFunction    user-defined callback function to call upon completion.
-   * @param[in] callbackData        user-defined data to pass to the `callbackFunction`.
-   *
-   * @returns The `shared_ptr<ucxx::RequestTag>` object
-   */
-  friend std::shared_ptr<RequestAm> createRequestAmRecv(
-    std::shared_ptr<Endpoint> endpoint,
-    const bool enablePythonFuture,
-    RequestCallbackUserFunction callbackFunction,
-    RequestCallbackUserData callbackData);
+  friend std::shared_ptr<RequestAm> createRequestAm(std::shared_ptr<Endpoint> endpoint,
+                                                    const data::RequestData requestData,
```

Comment:
Given we only store `data::RequestData` at the base class `ucxx::Request` we would then need to branch again at some point to extract which of the types was passed to convert it back to `data::RequestData`, as you suggest in https://github.com/rapidsai/ucxx/pull/121#discussion_r1422514425. The alternative here would be not store it at the base class anymore and make it a member of each class which might make sense here since we're using an interface. However, although it has been fun to poke at `std::visit` I admit I didn't see this earlier but we may have drifted to a different paradigm here and perhaps it would have made more sense earlier to stick to inheritance for consistency, thus a different alternative (to keep consistency) would be to use inheritance, e.g.:

```cpp
class RequestData {};

class RequestDataAm : public RequestData {};

class RequestDataAmSend : public RequestDataAm {
    RequestDataAmSend(void* buffer, size_t length) : RequestDataAm()
};
```

And instead of `std::visit` we could simply use conditionals:

```cpp
    if (auto *amSend = dynamic_cast<RequestDataAmSend *>(&requestData)) {
        ...
    } else if (auto *amReceive = dynamic_cast<RequestDataAmReceive *>(&requestData)) {
        ...
    } else {
        throw std::runtime_error("Unreachable");
    }
```

Constructors would then receive the base class for their types:

```cpp
createRequestAm(..., const RequestDataAm requestDataAm, ...)
{
    RequestAm(..., requestDataAm, ...);
}

RequestAm(const RequestDataAm requestDataAm) : Request(requestDataAm)
{
}
```

And the base `Request` class would then be able to store the data as it does now:

```cpp
Request::Request(..., const RequestData requestData) : _requestData(requestData)
```

Just to be clear, I'm not necessarily in favor nor against neither of the solutions, to me it's clear they both have their advantages and disadvantages from both an API and error handling. Ultimately, they will both accomplish the same task and I think they will have similar code length, for now I think we can keep what we have instead of reworking everything. I'm still curious at what you think, should we try keeping consistency (here that would mean inheritance instead of interface)? Can you see large advantages of using the current `std::visit` approach instead of inheritance?

Anyway done in https://github.com/rapidsai/ucxx/pull/121/commits/3558c0b6c42c4fe9f2a55cf2cff9882abe884c98 .